### ISSUE TITLE
Calcite number input: allow configuration to prevent entry of specific characters 

### ISSUE THREAD

**Issue by ade10176** [Esri/calcite-design-system#10406]
**monday.com sync:** #18005803087

### Check existing issues

- [X] I have [checked for existing issues](https://github.com/Esri/calcite-design-system/issues) to avoid duplicates

### Description

Calcite number input currently supports characters like `-` or `e`. In a min/max file attachment use case, where `-` is not an acceptable entry, user currently can type in `-` and is persisted when:
• there are no numbers in the input (`-`)
• following numbers in the input (`5-`)

if the user enters `5-1`, the `-` is omitted, and the entry becomes `51`

if the user enters `-5`, the entry is not accepted and it triggers a validation error

### Acceptance Criteria

~Allow configuration to prevent specific character(s) to be entered in the number input.~ Also welcome to workaround ideas.

To match native behavior, the Calcite Number Input won't prevent entering valid non-digit characters, but will instead trigger a validation error state when the cases called out in the description of this issue occur.  

A few examples for when an error state should occur that is not already occurring:
- Just `e` in the input
- Just `-` in the input
- Just `E` in the input
- Just `.` in the input
- Just `,` in the input
- `5-1`, `5-` or other combinations that don't compute to a valid number

Examples for when an error state should NOT occur
- A number ending in a decimal separator `5.` or `5,` (depending on locale) since these are valid number expressions since the decimal at the end is implied

### Relevant Info

Field Maps Designer issue https://github.com/ArcGIS/field-apps-web/issues/4229

### Which Component

calcite input number

### Priority impact

impact - p1 - need for current milestone

### Calcite package

- [X] @esri/calcite-components
- [ ] @esri/calcite-components-angular
- [ ] @esri/calcite-components-react
- [ ] @esri/calcite-design-tokens
- [ ] @esri/eslint-plugin-calcite-components

### Esri team

ArcGIS Field Apps

**Comment by eriklharper** [comment]
@ade10176 does the `-5` validation error happen when `min` is set to something higher than `-5`?  If so, that's correct behavior, and by design. 

Couple requests as well:  first, could you create a codepen that demonstrates your issues with the input-number component?  Second, that Field Maps link 404s for me, so if you could update that also it will help me with that extra context.  Thanks!

**Comment by ade10176** [comment]
Here's the codepen from @nwhittaker as an example @eriklharper 
https://codepen.io/nwhittaker-esri/pen/zxvEWZy. It includes a native <input> for comparison with what the expected behavior should probably be.

**Comment by eriklharper** [comment]
Thanks for the Codepen!  To clarify, would it meet your acceptance criteria for this issue, if on blur after entering an out of range number, it highlights the border in red and maybe also displays the validation error text like the native one does? 

<img width="329" height="136" alt="Image" src="https://github.com/user-attachments/assets/f5f31712-a2f5-4221-adfb-6f2552dbff0c" />

**Comment by nwhittaker** [comment]
@eriklharper even if reporting was happening earlier and more frequently (which would be great), it doesn't meet our acceptance criteria because there's still a bug here around what Calcite is considering an invalid entry.

For example, the Calcite input shows both `-` and `4-` as valid entries while the native input does not.

>…if on blur after entering an out of range number, it highlights the border in red and maybe also displays the validation error text like the native one does?

I like this but I'm not sure there's a one-size-fits-all approach. In some cases, we'd ideally like to report validity on either the `input` or `change` events. In other cases, we'd want to report immediately on load.

How about exposing a `reportValidity()` method on Calcite's form-associated components similar to the method of the same name on native `<input>` elements?

**Comment by eriklharper** [comment]
> ...it doesn't meet our acceptance criteria because there's still a bug here around what Calcite is considering an invalid entry.

What I proposed above when I mentioned "if on blur after entering an out of range number, it highlights the border in red and maybe also displays the validation error text like the native one does" would imply that the input would consider a negative number as invalid for the purposes of form validity.  Sounds like there's a bug with getting it to trigger an invalid status in this case, so that's where we'll start to address this issue.

Agree with you also that it would be nice to have the flexibility to report validity in all those different ways.  I'll look into that as well.

**Comment by amring1** [comment]
+1 from the analysis team

Curious - is there a reason to allow entering a negative sign after a number?

<img width="607" height="266" alt="Image" src="https://github.com/user-attachments/assets/afefd2d8-22c6-4f84-ae00-f98152ef950d" />

**Comment by eriklharper** [comment]
> +1 from the analysis team
> 
> Curious - is there a reason to allow entering a negative sign after a number?
> 
> <img alt="Image" width="607" height="266" src="https://private-user-images.githubusercontent.com/68953430/623366709-afefd2d8-22c6-4f84-ae00-f98152ef950d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODQzMjM2ODgsIm5iZiI6MTc4NDMyMzM4OCwicGF0aCI6Ii82ODk1MzQzMC82MjMzNjY3MDktYWZlZmQyZDgtMjJjNi00Zjg0LWFlMDAtZjk4MTUyZWY5NTBkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA3MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNzE3VDIxMjMwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3ZDA4MWJkMzM1ZjRmNmMxMDk5N2VkMTU0NDQ3OWI2ZTFlMzY2NWFlMzNmN2U3YjE3OGU1ZmIyMzQ2OGE4NDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.kxJYPT4Ty4TfNCQZiDBcH0TgPQsEp9BX0t_VbNAbtBk">

Yes, you need to be able to enter negative exponents like `-2e-4`.

**Comment by amring1** [comment]
> > Curious - is there a reason to allow entering a negative sign after a number?
> 
> Yes, you need to be able to enter negative exponents like `-2e-4`.

Shouldn't that only follow an `e` then? Not a digit?
